pkg/controller/service/with.go

```diff
@@ -1,250 +1,541 @@
-				w.WriteHeader(http.StatusInternalServerError)
-				return
-			}
-			tokenID = token.ID
-		}
+		f(user)
+	})
+}
+
+func (s *Service) withSuperUserAuth(w http.ResponseWriter, r *http.Request, user *models.User, f func()) {
+	if user == nil {
+		log.WithError(ErrDependencyNotSupplied).Error("verifying super-user auth")
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	if !user.SuperAdmin {
+		w.WriteHeader(http.StatusForbidden)
+		return
+	}
+
+	f()
+}
+
+func (s *Service) withServiceAccount(w http.ResponseWriter, r *http.Request, project *models.Project, f func(serviceAccount *models.ServiceAccount)) {
+	if project == nil {
+		log.WithError(ErrDependencyNotSupplied).Error("getting service account")
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	vars := mux.Vars(r)
+	serviceAccountIdentifier := vars["serviceaccount"]
+	if serviceAccountIdentifier == "" {
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	var serviceAccount *models.ServiceAccount
+	var err error
+	if strings.Contains(serviceAccountIdentifier, "_") {
+		serviceAccount, err = s.serviceAccounts.GetServiceAccount(r.Context(), serviceAccountIdentifier, project.ID)
+	} else {
+		serviceAccount, err = s.serviceAccounts.LookupServiceAccount(r.Context(), serviceAccountIdentifier, project.ID)
+	}
+	if err == store.ErrServiceAccountNotFound {
+		http.Error(w, err.Error(), http.StatusNotFound)
+		return
+	} else if err != nil {
+		log.WithError(err).Error("lookup service account")
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	f(serviceAccount)
+}
+
+func (s *Service) withApplication(w http.ResponseWriter, r *http.Request, project *models.Project, f func(application *models.Application)) {
+	if project == nil {
+		log.WithError(ErrDependencyNotSupplied).Error("getting application")
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	vars := mux.Vars(r)
+	applicationIdentifier := vars["application"]
+	if applicationIdentifier == "" {
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	var application *models.Application
+	var err error
+	if strings.Contains(applicationIdentifier, "_") {
+		application, err = s.applications.GetApplication(r.Context(), applicationIdentifier, project.ID)
+	} else {
+		application, err = s.applications.LookupApplication(r.Context(), applicationIdentifier, project.ID)
+	}
+	if err == store.ErrApplicationNotFound {
+		http.Error(w, err.Error(), http.StatusNotFound)
+		return
+	} else if err != nil {
+		log.WithError(err).Error("lookup application")
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	f(application)
+}
 
-		handler(w, r, projectID, authenticatedUserID, authenticatedServiceAccountID, tokenID)
+func (s *Service) withRelease(w http.ResponseWriter, r *http.Request, project *models.Project, application *models.Application, f func(release *models.Release)) {
+	if application == nil || project == nil {
```

Comment:
We could remove all these checks if we change `application *models.Application` -> `application models.Application`